### DISCUSSION TITLE
simple/<module>/json 404 not found when using pipliteUrls

### DISCUSSION THREAD

**Discussion by laynr** [jupyterlite/jupyterlite#520] [Q&A] [answered]
Trying to use jupyterlite `0.1.0-beta.0` offline with a mirror of pypi specified with `pipliteUrls`.  Getting 404 errors, and then a failure in get_pypi_json when jupterlite trys to pull https://`offline_pypi_mirror`/simple/`module`/json .  The whl files exists in https://`offline pypi mirror`/simple/`module`/ but there is no json file. The first failed module is `traitlets`

What I have done:

1. Built with ` jupyter lite server - pyodide https://github.com/pyodide/pyodide/release/download/0.19.1.tar.bz2`
2. Modified jupyter-lite.json by adding pipliteUrls[] and disablePyPiFallback
3. Host offline

The mirror of pypi.org is working for pip.   

- Am I dreaming that micropip will magically convert whls to pyolite compatible modules (wasm)?  
- What are these JSON files that are being looked for, am I meant to create them?
- Should I just pre-download these whl files, what directory should I put them in?
- Basically what am I doing wrong?

Thank you

**Comment by bollwyvl** [comment, marked as answer]
Yep, there are rather a few moving parts.

Pyodide's micropip, and by extension, piplite, only works with:

- specially constructed pairs of js/wam files, as distributed with pyodide
and listed on its packages.json
- py3, noarch .whl files from a warehouse json api (like pypi.org), or by
explicit url

To meet the needs of our packages, piplite will also use a lightly
compacted version of many warehouse responses in all.json... the simplest
of which is generated by putting .whl files in ./pypi before  "jupyter lite
build"

The "examples" directory in this repo has an alternate approach, where the
urls of the wheels are captured explicitly before the build, then cached at
build time.
